File: golang/authentication/did_transition.go
package authentication

import (
	"fmt"
	"regexp"
	"strings"
	"time"

	"github.com/agent-network-protocol/anp/golang/proof"
)

const (
	ANPDidSuperseded         = 1019
	ANPDidTransitionInvalid  = 1020
	ANPDidTransitionConflict = 1021
	DefaultMaxTransitionHops = 8
)

type TransitionAssurance string

const (
	TransitionAssuranceVerified         TransitionAssurance = "verified"
	TransitionAssuranceRecoveryVerified TransitionAssurance = "recovery_verified"
	TransitionAssuranceProviderAsserted TransitionAssurance = "provider_asserted"
	TransitionAssuranceUnverified       TransitionAssurance = "unverified"
)

type TransitionStatus string

const (
	TransitionStatusActive     TransitionStatus = "active"
	TransitionStatusSuperseded TransitionStatus = "superseded"
)

type TransitionErrorKind string

const (
	TransitionErrorUnsupportedProfile       TransitionErrorKind = "unsupported_profile"
	TransitionErrorInvalidDocument          TransitionErrorKind = "invalid_document"
	TransitionErrorInvalidProof             TransitionErrorKind = "invalid_proof"
	TransitionErrorRecoveryNotPreauthorized TransitionErrorKind = "recovery_not_preauthorized"
	TransitionErrorInvalidProviderAssertion TransitionErrorKind = "invalid_provider_assertion"
	TransitionErrorStablePathMismatch       TransitionErrorKind = "stable_path_mismatch"
	TransitionErrorDirectSuccessorRequired  TransitionErrorKind = "direct_successor_required"
	TransitionErrorCycle                    TransitionErrorKind = "cycle"
	TransitionErrorConflict                 TransitionErrorKind = "conflict"
	TransitionErrorMaxHopsExceeded          TransitionErrorKind = "max_hops_exceeded"
	TransitionErrorNetwork                  TransitionErrorKind = "network_error"
)

type DidTransitionError struct {
	Kind    TransitionErrorKind
	Code    int
	Message string
}

func (e *DidTransitionError) Error() string {
	return fmt.Sprintf("%s: %s", e.Kind, e.Message)
}

func newTransitionError(kind TransitionErrorKind, message string) *DidTransitionError {
	code := ANPDidTransitionInvalid
	if kind == TransitionErrorCycle || kind == TransitionErrorConflict || kind == TransitionErrorMaxHopsExceeded {
		code = ANPDidTransitionConflict
	}
	return &DidTransitionError{Kind: kind, Code: code, Message: message}
}

type DidWbaE1Profile struct {
	DID               string
	Origin            string
	StableSubjectPath string
	Fingerprint       string
}

type TransitionHop struct {
	PredecessorDID string              `json:"predecessor_did"`
	SuccessorDID   string              `json:"successor_did"`
	Assurance      TransitionAssurance `json:"assurance"`
}

type TransitionResult struct {
	RequestedDID string               `json:"requested_did"`
	CurrentDID   string               `json:"current_did"`
	Status       TransitionStatus     `json:"status"`
	Assurance    *TransitionAssurance `json:"assurance"`
	Hops         []TransitionHop      `json:"hops"`
}

type DidDocumentFetcher func(did string) (map[string]any, error)

type TransitionCache interface {
	GetSuccessor(predecessorDID string) (string, bool)
	CompareAndSet(predecessorDID string, successorDID string) bool
}

type InMemoryTransitionCache struct {
	edges map[string]string
}

func NewInMemoryTransitionCache(edges map[string]string) *InMemoryTransitionCache {
	copyEdges := make(map[string]string, len(edges))
	for predecessor, successor := range edges {
		copyEdges[predecessor] = successor
	}
	return &InMemoryTransitionCache{edges: copyEdges}
}

func (c *InMemoryTransitionCache) GetSuccessor(predecessorDID string) (string, bool) {
	successor, ok := c.edges[predecessorDID]
	return successor, ok
}

func (c *InMemoryTransitionCache) CompareAndSet(predecessorDID string, successorDID string) bool {
	if existing, ok := c.edges[predecessorDID]; ok {
		return existing == successorDID
	}
	c.edges[predecessorDID] = successorDID
	return true
}

var e1SegmentPattern = regexp.MustCompile(`^e1_[A-Za-z0-9_-]{43}$`)

func ParseDidWbaE1(did string) (DidWbaE1Profile, error) {
	parts := strings.Split(did, ":")
	if len(parts) < 5 || parts[0] != "did" || parts[1] != "wba" || !e1SegmentPattern.MatchString(parts[len(parts)-1]) {
		return DidWbaE1Profile{}, newTransitionError(
			TransitionErrorUnsupportedProfile,
			"automatic transition requires a path-based did:wba E1 DID",
		)
	}
	return DidWbaE1Profile{
		DID:               did,
		Origin:            parts[2],
		StableSubjectPath: strings.Join(parts[2:len(parts)-1], ":"),
		Fingerprint:       parts[len(parts)-1][3:],
	}, nil
}

func requireDocumentID(did string, document map[string]any) error {
	if stringValue(document["id"]) != did {
		return newTransitionError(TransitionErrorInvalidDocument, "DID Document id mismatch")
	}
	return nil
}

func bindingMethod(document map[string]any, profile DidWbaE1Profile) map[string]any {
	methods, _ := document["verificationMethod"].([]any)
	for _, rawMethod := range methods {
		method, ok := rawMethod.(map[string]any)
		if !ok {
			continue
		}
		publicKey, err := ExtractPublicKey(method)
		if err != nil {
			continue
		}
		fingerprint, err := ComputeMultikeyFingerprint(publicKey)
		if err == nil && fingerprint == profile.Fingerprint {
			return method
		}
	}
	return nil
}

func VerifyActiveE1Document(did string, document map[string]any) error {
	profile, err := ParseDidWbaE1(did)
	if err != nil {
		return err
	}
	if err := requireDocumentID(did, document); err != nil {
		return err
	}
	if deactivated, _ := document["deactivated"].(bool); deactivated {
		return newTransitionError(TransitionErrorInvalidDocument, "expected an active DID Document")
	}
	proofValue, ok := document["proof"].(map[string]any)
	if !ok {
		return newTransitionError(TransitionErrorInvalidProof, "active E1 proof is required")
	}
	methodID := stringValue(proofValue["verificationMethod"])
	if methodID == "" || !IsAssertionMethodAuthorized(document, methodID) {
		return newTransitionError(TransitionErrorInvalidProof, "binding key is not authorized")
	}
	method := FindVerificationMethod(document, methodID)
	if method == nil {
		return newTransitionError(TransitionErrorInvalidProof, "binding key is missing")
	}
	publicKey, err := ExtractPublicKey(method)
	if err != nil {
		return newTransitionError(TransitionErrorInvalidProof, "invalid binding key")
	}
	fingerprint, err := ComputeMultikeyFingerprint(publicKey)
	if err != nil || fingerprint != profile.Fingerprint {
		return newTransitionError(TransitionErrorInvalidProof, "binding fingerprint mismatch")
	}
	if !proof.VerifyW3CProof(document, publicKey, proof.VerificationOptions{ExpectedPurpose: "assertionMethod"}) {
		return newTransitionError(TransitionErrorInvalidProof, "active E1 proof verification failed")
	}
	return nil
}

func verifyProviderAssertion(
	predecessorDocument map[string]any,
	predecessor DidWbaE1Profile,
	successor DidWbaE1Profile,
	providerFetcher DidDocumentFetcher,
) error {
	assertion, ok := predecessorDocument["providerTransitionAssertion"].(map[string]any)
	if !ok || len(assertion) != 7 {
		return newTransitionError(TransitionErrorInvalidProviderAssertion, "providerTransitionAssertion has an invalid shape")
	}
	for _, field := range []string{"type", "providerDid", "predecessorDid", "successorDid", "stableSubjectPath", "issuedAt", "proof"} {
		if _, exists := assertion[field]; !exists {
			return newTransitionError(TransitionErrorInvalidProviderAssertion, "providerTransitionAssertion has an invalid shape")
		}
	}
	providerDID := "did:wba:" + predecessor.Origin
	issuedAt := stringValue(assertion["issuedAt"])
	parsedTime, timeErr := time.Parse("2006-01-02T15:04:05Z", issuedAt)
	assertionProof, proofOK := assertion["proof"].(map[string]any)
	if stringValue(assertion["type"]) != "DidWbaProviderTransitionAssertion" ||
		stringValue(assertion["providerDid"]) != providerDID ||
		stringValue(assertion["predecessorDid"]) != predecessor.DID ||
		stringValue(assertion["successorDid"]) != successor.DID ||
		stringValue(assertion["stableSubjectPath"]) != predecessor.StableSubjectPath ||
		timeErr != nil || parsedTime.UTC().Format("2006-01-02T15:04:05Z") != issuedAt ||
		!proofOK || stringValue(assertionProof["created"]) != issuedAt {
		return newTransitionError(TransitionErrorInvalidProviderAssertion, "providerTransitionAssertion binding fields are invalid")
	}
	if providerFetcher == nil {
		return newTransitionError(TransitionErrorInvalidProviderAssertion, "provider DID resolver is required")
	}
	providerDocument, err := providerFetcher(providerDID)
	if err != nil {
		return newTransitionError(TransitionErrorNetwork, err.Error())
	}
	if err := requireDocumentID(providerDID, providerDocument); err != nil {
		return err
	}
	methodID := stringValue(assertionProof["verificationMethod"])
	if !strings.HasPrefix(methodID, providerDID+"#") || !IsAssertionMethodAuthorized(providerDocument, methodID) {
		return newTransitionError(TransitionErrorInvalidProviderAssertion, "provider proof key is not authorized")
	}
	method := FindVerificationMethod(providerDocument, methodID)
	if method == nil {
		return newTransitionError(TransitionErrorInvalidProviderAssertion, "provider proof key is missing")
	}
	publicKey, err := ExtractPublicKey(method)
	if err != nil || !proof.VerifyW3CProof(assertion, publicKey, proof.VerificationOptions{ExpectedPurpose: "assertionMethod"}) {
		return newTransitionError(TransitionErrorInvalidProviderAssertion, "providerTransitionAssertion proof verification failed")
	}
	return nil
}

func VerifyTransitionHop(
	predecessorDocument map[string]any,
	successorDocument map[string]any,
	trustedPredecessor map[string]any,
	providerFetcher DidDocumentFetcher,
) (TransitionHop, error) {
	predecessorDID := stringValue(predecessorDocument["id"])
	successorDID := stringValue(predecessorDocument["successorDid"])
	if predecessorDID == "" || successorDID == "" {
		return TransitionHop{}, newTransitionError(TransitionErrorInvalidDocument, "deactivated predecessor requires id and successorDid")
	}
	predecessor, err := ParseDidWbaE1(predecessorDID)
	if err != nil {
		return TransitionHop{}, err
	}
	successor, err := ParseDidWbaE1(successorDID)
	if err != nil {
		return TransitionHop{}, err
	}
	if err := requireDocumentID(successorDID, successorDocument); err != nil {
		return TransitionHop{}, err
	}
	deactivated, _ := predecessorDocument["deactivated"].(bool)
	if !deactivated {
		return TransitionHop{}, newTransitionError(TransitionErrorInvalidDocument, "predecessor is not deactivated")
	}
	if predecessor.StableSubjectPath != successor.StableSubjectPath {
		return TransitionHop{}, newTransitionError(TransitionErrorStablePathMismatch, "predecessor and successor stable subject paths differ")
	}
	binding := bindingMethod(predecessorDocument, predecessor)
	if binding == nil {
		return TransitionHop{}, newTransitionError(TransitionErrorInvalidDocument, "deactivated predecessor does not retain its binding key")
	}
	if aliases, ok := successorDocument["alsoKnownAs"].([]any); ok {
		matching := make([]string, 0)
		for _, rawAlias := range aliases {
			alias, ok := rawAlias.(string)
			if !ok {
				continue
			}
			aliasProfile, parseErr := ParseDidWbaE1(alias)
			if parseErr == nil && aliasProfile.StableSubjectPath == successor.StableSubjectPath {
				matching = append(matching, alias)
			}
		}
		if len(matching) > 0 {
			found := false
			for _, alias := range matching {
				found = found || alias == predecessorDID
			}
			if !found {
				return TransitionHop{}, newTransitionError(TransitionErrorDirectSuccessorRequired, "successor identifies a different direct predecessor")
			}
		}
	}

	_, providerPresent := predecessorDocument["providerTransitionAssertion"]
	if providerPresent {
		if err := verifyProviderAssertion(predecessorDocument, predecessor, successor, providerFetcher); err != nil {
			return TransitionHop{}, err
		}
	}

	assurance := TransitionAssuranceUnverified
	if rawProof, proofPresent := predecessorDocument["proof"]; proofPresent {
		transitionProof, ok := rawProof.(map[string]any)
		if !ok {
			return TransitionHop{}, newTransitionError(TransitionErrorInvalidProof, "transition proof is invalid")
		}
		methodID := stringValue(transitionProof["verificationMethod"])
		if stringValue(binding["id"]) == methodID {
			publicKey, extractErr := ExtractPublicKey(binding)
			if extractErr != nil || !proof.VerifyW3CProof(predecessorDocument, publicKey, proof.VerificationOptions{ExpectedPurpose: "assertionMethod"}) {
				return TransitionHop{}, newTransitionError(TransitionErrorInvalidProof, "old binding proof failed")
			}
			assurance = TransitionAssuranceVerified
		} else {
			trustedMethod := FindVerificationMethod(trustedPredecessor, methodID)
			if trustedPredecessor == nil || trustedMethod == nil || !IsAssertionMethodAuthorized(trustedPredecessor, methodID) {
				return TransitionHop{}, newTransitionError(TransitionErrorRecoveryNotPreauthorized, "recovery key was not pre-authorized by the trusted predecessor")
			}
			publicKey, extractErr := ExtractPublicKey(trustedMethod)
			if extractErr != nil || !proof.VerifyW3CProof(predecessorDocument, publicKey, proof.VerificationOptions{ExpectedPurpose: "assertionMethod"}) {
				return TransitionHop{}, newTransitionError(TransitionErrorInvalidProof, "recovery proof failed")
			}
			assurance = TransitionAssuranceRecoveryVerified
		}
	} else if providerPresent {
		assurance = TransitionAssuranceProviderAsserted
	}
	return TransitionHop{PredecessorDID: predecessorDID, SuccessorDID: successorDID, Assurance: assurance}, nil
}

func assuranceRank(value TransitionAssurance) int {
	switch value {
	case TransitionAssuranceVerified:
		return 0
	case TransitionAssuranceRecoveryVerified:
		return 1
	case TransitionAssuranceProviderAsserted:
		return 2
	default:
		return 3
	}
}

func ResolveCurrentDID(
	requestedDID string,
	fetcher DidDocumentFetcher,
	trustedDocuments map[string]map[string]any,
	providerFetcher DidDocumentFetcher,
	cache TransitionCache,
	maxHops int,
) (TransitionResult, error) {
	if _, err := ParseDidWbaE1(requestedDID); err != nil {
		return TransitionResult{}, err
	}
	if maxHops < 1 {
		return TransitionResult{}, newTransitionError(TransitionErrorMaxHopsExceeded, "maxHops must be positive")
	}
	current := requestedDID
	visited := make(map[string]bool)
	hops := make([]TransitionHop, 0)
	for {
		if visited[current] {
			return TransitionResult{}, newTransitionError(TransitionErrorCycle, "transition chain contains a cycle")
		}
		visited[current] = true
		document, err := fetcher(current)
		if err != nil {
			return TransitionResult{}, newTransitionError(TransitionErrorNetwork, err.Error())
		}
		if err := requireDocumentID(current, document); err != nil {
			return TransitionResult{}, err
		}
		deactivated, _ := document["deactivated"].(bool)
		if !deactivated {
			if err := VerifyActiveE1Document(current, document); err != nil {
				return TransitionResult{}, err
			}
			var assurance *TransitionAssurance
			for _, hop := range hops {
				if assurance == nil || assuranceRank(hop.Assurance) > assuranceRank(*assurance) {
					value := hop.Assurance
					assurance = &value
				}
			}
			status := TransitionStatusActive
			if len(hops) > 0 {
				status = TransitionStatusSuperseded
			}
			return TransitionResult{RequestedDID: requestedDID, CurrentDID: current, Status: status, Assurance: assurance, Hops: hops}, nil
		}
		if len(hops) >= maxHops {
			return TransitionResult{}, newTransitionError(TransitionErrorMaxHopsExceeded, "transition chain exceeds maxHops")
		}
		successorDID := stringValue(document["successorDid"])
		if successorDID == "" {
			return TransitionResult{}, newTransitionError(TransitionErrorInvalidDocument, "deactivated transition document has no successorDid")
		}
		if visited[successorDID] {
			return TransitionResult{}, newTransitionError(TransitionErrorCycle, "transition chain contains a cycle")
		}
		successorDocument, fetchErr := fetcher(successorDID)
		if fetchErr != nil {
			return TransitionResult{}, newTransitionError(TransitionErrorNetwork, fetchErr.Error())
		}
		hop, verifyErr := VerifyTransitionHop(document, successorDocument, trustedDocuments[current], providerFetcher)
		if verifyErr != nil {
			return TransitionResult{}, verifyErr
		}
		if !cache.CompareAndSet(current, successorDID) {
			return TransitionResult{}, newTransitionError(TransitionErrorConflict, "a different successor is already cached for predecessor")
		}
		hops = append(hops, hop)
		current = successorDID
	}
}

File: anp/authentication/did_transition.py
"""Verification and forward resolution for did:wba E1 transitions."""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Callable, Dict, Mapping, Optional, Protocol

from cryptography.hazmat.primitives.asymmetric import ed25519

from anp.proof import verify_w3c_proof

from .did_wba import (
    _extract_public_key,
    _find_verification_method,
    _is_assertion_method_authorized_in_document,
    compute_multikey_fingerprint,
)


ANP_DID_SUPERSEDED = 1019
ANP_DID_TRANSITION_INVALID = 1020
ANP_DID_TRANSITION_CONFLICT = 1021
DEFAULT_MAX_TRANSITION_HOPS = 8

_E1_SEGMENT = re.compile(r"^e1_[A-Za-z0-9_-]{43}$")
_CANONICAL_UTC = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$")
_PROVIDER_FIELDS = {
    "type",
    "providerDid",
    "predecessorDid",
    "successorDid",
    "stableSubjectPath",
    "issuedAt",
    "proof",
}


class TransitionAssurance(str, Enum):
    VERIFIED = "verified"
    RECOVERY_VERIFIED = "recovery_verified"
    PROVIDER_ASSERTED = "provider_asserted"
    UNVERIFIED = "unverified"


class TransitionStatus(str, Enum):
    ACTIVE = "active"
    SUPERSEDED = "superseded"


class TransitionErrorKind(str, Enum):
    UNSUPPORTED_PROFILE = "unsupported_profile"
    INVALID_DOCUMENT = "invalid_document"
    INVALID_PROOF = "invalid_proof"
    RECOVERY_NOT_PREAUTHORIZED = "recovery_not_preauthorized"
    INVALID_PROVIDER_ASSERTION = "invalid_provider_assertion"
    STABLE_PATH_MISMATCH = "stable_path_mismatch"
    DIRECT_SUCCESSOR_REQUIRED = "direct_successor_required"
    CYCLE = "cycle"
    CONFLICT = "conflict"
    MAX_HOPS_EXCEEDED = "max_hops_exceeded"
    NETWORK_ERROR = "network_error"


_CONFLICT_ERRORS = {
    TransitionErrorKind.CYCLE,
    TransitionErrorKind.CONFLICT,
    TransitionErrorKind.MAX_HOPS_EXCEEDED,
}


class DidTransitionError(ValueError):
    def __init__(self, kind: TransitionErrorKind, message: str):
        self.kind = kind
        self.code = (
            ANP_DID_TRANSITION_CONFLICT
            if kind in _CONFLICT_ERRORS
            else ANP_DID_TRANSITION_INVALID
        )
        super().__init__(message)


@dataclass(frozen=True)
class DidWbaE1Profile:
    did: str
    origin: str
    stable_subject_path: str
    fingerprint: str


@dataclass(frozen=True)
class TransitionHop:
    predecessor_did: str
    successor_did: str
    assurance: TransitionAssurance


@dataclass(frozen=True)
class TransitionResult:
    requested_did: str
    current_did: str
    status: TransitionStatus
    assurance: Optional[TransitionAssurance]
    hops: tuple[TransitionHop, ...] = field(default_factory=tuple)


class TransitionCache(Protocol):
    def get_successor(self, predecessor_did: str) -> Optional[str]: ...

    def compare_and_set(self, predecessor_did: str, successor_did: str) -> bool: ...


class InMemoryTransitionCache:
    """Small CAS cache suitable for callers that do not have a persistent store."""

    def __init__(self, edges: Optional[Mapping[str, str]] = None):
        self._edges = dict(edges or {})

    def get_successor(self, predecessor_did: str) -> Optional[str]:
        return self._edges.get(predecessor_did)

    def compare_and_set(self, predecessor_did: str, successor_did: str) -> bool:
        existing = self._edges.get(predecessor_did)
        if existing is not None:
            return existing == successor_did
        self._edges[predecessor_did] = successor_did
        return True

    def snapshot(self) -> Dict[str, str]:
        return dict(self._edges)


DocumentFetcher = Callable[[str], Dict[str, Any]]


def parse_did_wba_e1(did: str) -> DidWbaE1Profile:
    parts = did.split(":")
    if len(parts) < 5 or parts[0:2] != ["did", "wba"]:
        raise DidTransitionError(
            TransitionErrorKind.UNSUPPORTED_PROFILE,
            "automatic transition requires a path-based did:wba E1 DID",
        )
    segment = parts[-1]
    if not _E1_SEGMENT.fullmatch(segment):
        raise DidTransitionError(
            TransitionErrorKind.UNSUPPORTED_PROFILE,
            "automatic transition supports only the e1 profile",
        )
    return DidWbaE1Profile(
        did=did,
        origin=parts[2],
        stable_subject_path=":".join(parts[2:-1]),
        fingerprint=segment[3:],
    )


def _binding_method(document: Dict[str, Any], profile: DidWbaE1Profile) -> Optional[dict]:
    for method in document.get("verificationMethod", []):
        if not isinstance(method, dict):
            continue
        try:
            public_key = _extract_public_key(method)
        except ValueError:
            continue
        if isinstance(public_key, ed25519.Ed25519PublicKey):
            if compute_multikey_fingerprint(public_key) == profile.fingerprint:
                return method
    return None


def _require_document_id(did: str, document: Dict[str, Any]) -> None:
    if document.get("id") != did:
        raise DidTransitionError(
            TransitionErrorKind.INVALID_DOCUMENT,
            f"DID Document id does not match {did}",
        )


def verify_active_e1_document(did: str, document: Dict[str, Any]) -> None:
    profile = parse_did_wba_e1(did)
    _require_document_id(did, document)
    if document.get("deactivated") is True:
        raise DidTransitionError(
            TransitionErrorKind.INVALID_DOCUMENT,
            "expected an active DID Document",
        )
    proof = document.get("proof")
    if not isinstance(proof, dict):
        raise DidTransitionError(TransitionErrorKind.INVALID_PROOF, "active E1 proof is required")
    verification_method_id = proof.get("verificationMethod")
    if not isinstance(verification_method_id, str):
        raise DidTransitionError(TransitionErrorKind.INVALID_PROOF, "proof verificationMethod is required")
    method = _find_verification_method(document, verification_method_id)
    if not method or not _is_assertion_method_authorized_in_document(
        document, verification_method_id
    ):
        raise DidTransitionError(TransitionErrorKind.INVALID_PROOF, "binding key is not authorized")
    try:
        public_key = _extract_public_key(method)
    except ValueError as exc:
        raise DidTransitionError(TransitionErrorKind.INVALID_PROOF, "invalid binding key") from exc
    if not isinstance(public_key, ed25519.Ed25519PublicKey):
        raise DidTransitionError(TransitionErrorKind.INVALID_PROOF, "binding key must be Ed25519")
    if compute_multikey_fingerprint(public_key) != profile.fingerprint:
        raise DidTransitionError(TransitionErrorKind.INVALID_PROOF, "binding fingerprint mismatch")
    if not verify_w3c_proof(document, public_key, expected_purpose="assertionMethod"):
        raise DidTransitionError(TransitionErrorKind.INVALID_PROOF, "active E1 proof verification failed")


def _provider_did(profile: DidWbaE1Profile) -> str:
    return f"did:wba:{profile.origin}"


def _verify_provider_assertion(
    predecessor_document: Dict[str, Any],
    predecessor: DidWbaE1Profile,
    successor: DidWbaE1Profile,
    provider_fetcher: Optional[DocumentFetcher],
) -> None:
    assertion = predecessor_document.get("providerTransitionAssertion")
    if not isinstance(assertion, dict) or set(assertion) != _PROVIDER_FIELDS:
        raise DidTransitionError(
            TransitionErrorKind.INVALID_PROVIDER_ASSERTION,
            "providerTransitionAssertion has an invalid shape",
        )
    provider_did = _provider_did(predecessor)
    issued_at = assertion.get("issuedAt")
    proof = assertion.get("proof")
    if (
        assertion.get("type") != "DidWbaProviderTransitionAssertion"
        or assertion.get("providerDid") != provider_did
        or assertion.get("predecessorDid") != predecessor.did
        or assertion.get("successorDid") != successor.did
        or assertion.get("stableSubjectPath") != predecessor.stable_subject_path
        or not isinstance(issued_at, str)
        or not _CANONICAL_UTC.fullmatch(issued_at)
        or not isinstance(proof, dict)
        or proof.get("created") != issued_at
    ):
        raise DidTransitionError(
            TransitionErrorKind.INVALID_PROVIDER_ASSERTION,
            "providerTransitionAssertion binding fields are invalid",
        )
    if provider_fetcher is None:
        raise DidTransitionError(
            TransitionErrorKind.INVALID_PROVIDER_ASSERTION,
            "provider DID resolver is required",
        )
    try:
        provider_document = provider_fetcher(provider_did)
    except Exception as exc:
        raise DidTransitionError(TransitionErrorKind.NETWORK_ERROR, str(exc)) from exc
    _require_document_id(provider_did, provider_document)
    method_id = proof.get("verificationMethod")
    if not isinstance(method_id, str) or not method_id.startswith(f"{provider_did}#"):
        raise DidTransitionError(
            TransitionErrorKind.INVALID_PROVIDER_ASSERTION,
            "provider proof key is outside providerDid",
        )
    method = _find_verification_method(provider_document, method_id)
    if not method or not _is_assertion_method_authorized_in_document(provider_document, method_id):
        raise DidTransitionError(
            TransitionErrorKind.INVALID_PROVIDER_ASSERTION,
            "provider proof key is not authorized",
        )
    try:
        public_key = _extract_public_key(method)
    except ValueError as exc:
        raise DidTransitionError(
            TransitionErrorKind.INVALID_PROVIDER_ASSERTION, "invalid provider proof key"
        ) from exc
    if not verify_w3c_proof(assertion, public_key, expected_purpose="assertionMethod"):
        raise DidTransitionError(
            TransitionErrorKind.INVALID_PROVIDER_ASSERTION,
            "providerTransitionAssertion proof verification failed",
        )


def verify_transition_hop(
    predecessor_document: Dict[str, Any],
    successor_document: Dict[str, Any],
    *,
    trusted_predecessor: Optional[Dict[str, Any]] = None,
    provider_fetcher: Optional[DocumentFetcher] = None,
) -> TransitionHop:
    predecessor_did = predecessor_document.get("id")
    successor_did = predecessor_document.get("successorDid")
    if not isinstance(predecessor_did, str) or not isinstance(successor_did, str):
        raise DidTransitionError(
            TransitionErrorKind.INVALID_DOCUMENT,
            "deactivated predecessor requires id and successorDid",
        )
    predecessor = parse_did_wba_e1(predecessor_did)
    successor = parse_did_wba_e1(successor_did)
    _require_document_id(successor_did, successor_document)
    if predecessor_document.get("deactivated") is not True:
        raise DidTransitionError(TransitionErrorKind.INVALID_DOCUMENT, "predecessor is not deactivated")
    if predecessor.stable_subject_path != successor.stable_subject_path:
        raise DidTransitionError(
            TransitionErrorKind.STABLE_PATH_MISMATCH,
            "predecessor and successor stable subject paths differ",
        )
    if _binding_method(predecessor_document, predecessor) is None:
        raise DidTransitionError(
            TransitionErrorKind.INVALID_DOCUMENT,
            "deactivated predecessor does not retain its binding key",
        )
    aliases = successor_document.get("alsoKnownAs")
    if isinstance(aliases, list):
        same_subject_predecessors = []
        for alias in aliases:
            if not isinstance(alias, str):
                continue
            try:
                alias_profile = parse_did_wba_e1(alias)
            except DidTransitionError:
                continue
            if alias_profile.stable_subject_path == successor.stable_subject_path:
                same_subject_predecessors.append(alias)
        if same_subject_predecessors and predecessor_did not in same_subject_predecessors:
            raise DidTransitionError(
                TransitionErrorKind.DIRECT_SUCCESSOR_REQUIRED,
                "successor identifies a different direct predecessor",
            )

    provider_present = "providerTransitionAssertion" in predecessor_document
    if provider_present:
        _verify_provider_assertion(
            predecessor_document, predecessor, successor, provider_fetcher
        )

    proof = predecessor_document.get("proof")
    assurance = TransitionAssurance.UNVERIFIED
    if proof is not None:
        if not isinstance(proof, dict):
            raise DidTransitionError(TransitionErrorKind.INVALID_PROOF, "transition proof is invalid")
        method_id = proof.get("verificationMethod")
        if not isinstance(method_id, str):
            raise DidTransitionError(TransitionErrorKind.INVALID_PROOF, "transition proof key is missing")
        binding = _binding_method(predecessor_document, predecessor)
        if binding and binding.get("id") == method_id:
            public_key = _extract_public_key(binding)
            if not verify_w3c_proof(
                predecessor_document, public_key, expected_purpose="assertionMethod"
            ):
                raise DidTransitionError(TransitionErrorKind.INVALID_PROOF, "old binding proof failed")
            assurance = TransitionAssurance.VERIFIED
        else:
            trusted_method = (
                _find_verification_method(trusted_predecessor, method_id)
                if trusted_predecessor
                else None
            )
            if not trusted_predecessor or not trusted_method or not _is_assertion_method_authorized_in_document(
                trusted_predecessor, method_id
            ):
                raise DidTransitionError(
                    TransitionErrorKind.RECOVERY_NOT_PREAUTHORIZED,
                    "recovery key was not pre-authorized by the trusted predecessor",
                )
            try:
                public_key = _extract_public_key(trusted_method)
            except ValueError as exc:
                raise DidTransitionError(TransitionErrorKind.INVALID_PROOF, "invalid recovery key") from exc
            if not verify_w3c_proof(
                predecessor_document, public_key, expected_purpose="assertionMethod"
            ):
                raise DidTransitionError(TransitionErrorKind.INVALID_PROOF, "recovery proof failed")
            assurance = TransitionAssurance.RECOVERY_VERIFIED
    elif provider_present:
        assurance = TransitionAssurance.PROVIDER_ASSERTED

    return TransitionHop(predecessor_did, successor_did, assurance)


_ASSURANCE_RANK = {
    TransitionAssurance.VERIFIED: 0,
    TransitionAssurance.RECOVERY_VERIFIED: 1,
    TransitionAssurance.PROVIDER_ASSERTED: 2,
    TransitionAssurance.UNVERIFIED: 3,
}


def resolve_current_did(
    requested_did: str,
    fetcher: DocumentFetcher,
    *,
    trusted_predecessor: Optional[Dict[str, Any]] = None,
    trusted_documents: Optional[Mapping[str, Dict[str, Any]]] = None,
    provider_fetcher: Optional[DocumentFetcher] = None,
    cache: Optional[TransitionCache] = None,
    max_hops: int = DEFAULT_MAX_TRANSITION_HOPS,
) -> TransitionResult:
    parse_did_wba_e1(requested_did)
    if max_hops < 1:
        raise DidTransitionError(TransitionErrorKind.MAX_HOPS_EXCEEDED, "max_hops must be positive")
    if cache is None:
        cache = InMemoryTransitionCache()
    trusted = dict(trusted_documents or {})
    if trusted_predecessor is not None:
        trusted[requested_did] = trusted_predecessor
    current = requested_did
    visited: set[str] = set()
    hops: list[TransitionHop] = []
    verified_edges: list[tuple[str, str]] = []

    while True:
        if current in visited:
            raise DidTransitionError(TransitionErrorKind.CYCLE, "transition chain contains a cycle")
        visited.add(current)
        try:
            document = fetcher(current)
        except DidTransitionError:
            raise
        except Exception as exc:
            raise DidTransitionError(TransitionErrorKind.NETWORK_ERROR, str(exc)) from exc
        _require_document_id(current, document)
        if document.get("deactivated") is not True:
            verify_active_e1_document(current, document)
            for predecessor_did, successor_did in verified_edges:
                if not cache.compare_and_set(predecessor_did, successor_did):
                    raise DidTransitionError(
                        TransitionErrorKind.CONFLICT,
                        "a different successor is already cached for predecessor",
                    )
            assurance = (
                max((hop.assurance for hop in hops), key=_ASSURANCE_RANK.get)
                if hops
                else None
            )
            return TransitionResult(
                requested_did=requested_did,
                current_did=current,
                status=TransitionStatus.SUPERSEDED if hops else TransitionStatus.ACTIVE,
                assurance=assurance,
                hops=tuple(hops),
            )
        if len(hops) >= max_hops:
            raise DidTransitionError(
                TransitionErrorKind.MAX_HOPS_EXCEEDED,
                "transition chain exceeds max_hops",
            )
        successor_did = document.get("successorDid")
        if not isinstance(successor_did, str):
            raise DidTransitionError(
                TransitionErrorKind.INVALID_DOCUMENT,
                "deactivated transition document has no successorDid",
            )
        if successor_did in visited:
            raise DidTransitionError(TransitionErrorKind.CYCLE, "transition chain contains a cycle")
        try:
            successor_document = fetcher(successor_did)
        except DidTransitionError:
            raise
        except Exception as exc:
            raise DidTransitionError(TransitionErrorKind.NETWORK_ERROR, str(exc)) from exc
        hop = verify_transition_hop(
            document,
            successor_document,
            trusted_predecessor=trusted.get(current),
            provider_fetcher=provider_fetcher,
        )
        cached_successor = cache.get_successor(current)
        if cached_successor is not None and cached_successor != successor_did:
            raise DidTransitionError(
                TransitionErrorKind.CONFLICT,
                "a different successor is already cached for predecessor",
            )
        if hop.assurance in {
            TransitionAssurance.VERIFIED,
            TransitionAssurance.RECOVERY_VERIFIED,
        }:
            verified_edges.append((current, successor_did))
        hops.append(hop)
        current = successor_did
